### PR TITLE
feat(nx-monorepo): add support for Python as an NX Project

### PR DESCRIPTION
**Pull request by agdimech**
PR [1/3]

This PR adds the capability for NX based monorepos to be completely written in python. In order to support this, some changes to TypeSafeApi were needed to resolve some invalid assumptions regarding parent monorepos being TypescriptBased.

2 additional PR's are planned as follows:

PR#2: Add Java support
PR#3: Refactor common code and add documentation and examples.

### CODE REVIEW

File: packages/type-safe-api/src/project/type-safe-api-project.ts [hunk 1]
```diff
@@ -256,16 +256,16 @@ export class TypeSafeApiProject extends Project {
...
       : libraryDir;
 
     // Declare the generated runtime projects
     const generatedLibraryProjects = generateLibraryProjects(libraries, {
-      parent: parentMonorepo ?? this,
+      parent: nxWorkspace ? this.parent! : this,
       parentPackageName: this.name,
       generatedCodeDir: libraryDirRelativeToParent,
-      isWithinMonorepo: !!parentMonorepo,
+      isWithinMonorepo: !!this.parent,
```
Comment (cogwirrel) [inline packages/type-safe-api/src/project/type-safe-api-project.ts]: This should be `!!nxWorkspace` :)

File: packages/type-safe-api/src/project/type-safe-api-project.ts [hunk 2]
```diff
@@ -397,9 +394,9 @@ export class TypeSafeApiProject extends Project {
     }
     this.infrastructure = infraProjects;
 
-    parentMonorepo?.addImplicitDependency?.(infraProject, this.model);
+    NxProject.ensure(infraProject).addImplicitDependency(this.model);
 
-    if (!parentMonorepo) {
+    if (!this.parent) {
```
Comment (cogwirrel) [inline packages/type-safe-api/src/project/type-safe-api-project.ts]: This should also be `!nxWorkspace`

File: packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+  constructor(options: NxMonorepoPythonProjectOptions) {
+    super({
+      ...options,
+      sample: false,
+      poetry: true,
+      pytest: false,
+    });
+
+    this.installTask = this.tasks.tryFind("install");
+    this.addDevDependency("aws-prototyping-sdk-nx-monorepo@0.0.0");
```
Comment (cogwirrel) [inline packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts]: Don't forget to name this the same as what's published to pip ;)

File: packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  synth() {
+    this.validateSubProjects();
+    this.setupPythonNx();
+    this.installPythonSubprojects();
+
+    super.synth();
+  }
+
+  postSynthesize(): void {
+    const vEnv = process.env.VIRTUAL_ENV;
```
Comment (cogwirrel) [inline packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts:237]: Probably worth a comment to explain our haxxx haha 😄

File: packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+          dependsOn: ["^install"],
+        });
+      });
+    }
+  }
+
+  /**
+   * Ensures subprojects don't have a default task and that all Python subpackages are configured to use Poetry.
+   */
+  private validateSubProjects() {
```
Comment (cogwirrel) [inline packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts:285]: Might want to update the PDKPipeline python project to use poetry :) don't mind if that's a separate change though :)

File: packages/nx-monorepo/src/projects/typescript/nx-monorepo.ts [hunk 6]
```diff
@@ -133,7 +133,7 @@ interface OverrideNxBuildTaskOptions {
  * This project type will bootstrap a NX based monorepo with support for polygot
  * builds, build caching, dependency graph visualization and much more.
  *
- * @pjid nx-monorepo
+ * @pjid nx-monorepo-ts
```
Comment (cogwirrel) [inline packages/nx-monorepo/src/projects/typescript/nx-monorepo.ts:136]: Might need to consider this a breaking change? If you can have multiple `pjid`s maybe we can also keep `nx-monorepo`?

File: packages/type-safe-api/src/project/type-safe-api-project.ts (hunk 1 above)
Comment (cogwirrel) [reply packages/type-safe-api/src/project/type-safe-api-project.ts]: Will need to update this here and for `generateInfraProject` below as this tells the generated typescript projects whether or not to run the `yarn/pnpm/npm link` commands.

Thinking about it though, I wonder if it actually needs to be `isWithinTsMonorepo` as I suppose with a Python based monorepo we don't have an overarching yarn/pnpm/npm workspace so the packages would still need to be linked in order to see one another?

**Review by cogwirrel — APPROVED** [review]

File: packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts (hunk 5 above)
Comment (agdimech) [reply packages/nx-monorepo/src/projects/python/nx-monorepo-py.ts:285]: will do.

File: packages/nx-monorepo/src/projects/typescript/nx-monorepo.ts (hunk 6 above)
Comment (agdimech) [reply packages/nx-monorepo/src/projects/typescript/nx-monorepo.ts:136]: its only used on initial bootstrap so not a breaking change technically. I will update docs in a seperate PR to mention `npx projen new --from @aws-prototyping-sdk/nx-monorepo nx-monorepo-ts` is the new way to bootstrap

**Comment by agdimech** [comment]
latest rev replaces all isWithinMonorepo instances to `this.parent instanceof NxMonorepoProject`

**Review by cogwirrel — APPROVED** [review]

**Comment by cogwirrel** [comment]
Great stuff! Awesome to have fully-fledged python support now! 😄

File: packages/type-safe-api/src/project/type-safe-api-project.ts (hunk 1 above)
Comment (agdimech) [reply packages/type-safe-api/src/project/type-safe-api-project.ts]: Updated in latest revision

**Pull request merged**